src/shiv/builder.py

```diff
@@ -46,6 +50,34 @@ def write_file_prefix(f: IO[Any], interpreter: str) -> None:
     f.write(b"#!" + interpreter.encode(sys.getfilesystemencoding()) + b"\n")
 
 
+def _write_to_zip_app(
+    zfile: zipfile.ZipFile, arcname: str, data_source: Union[Path, bytes],
+    date_time: Tuple[int, int, int, int, int, int], compression: int,
+    contents_hash: Optional[hashlib.sha3_256] = None,
+) -> None:
+    """Write a file or a bytestring to a ZipFile as a separate entry.
+
+    The approach is borrowed from 'wheel' code.
+    """
+    if isinstance(data_source, Path):
+        with data_source.open('rb') as f:
+            data = f.read()
+            st: Optional[os.stat_result] = os.fstat(f.fileno())
+    else:
+        data = data_source
+        st = None
```

Comment:
We can ditch the conditional, and since we know we are always passing in a `Path` object we can simplify it to:

    data = path.read_bytes()
    st = path.stat()

Reply:
We can still have `bytes` here when we write `environment.json` and `__main__.py` files.

Reply:
ah, good point! perhaps we should just make the argument "data" and do the `path.read_bytes()` call in the calling function?

Reply:
Then we won't be able to copy the file permissions. 
Actually, I think we shouldn't copy them anyway because it will make archive unreproducible on systems with different umask.

Reply:
In `TestCLI.test_hello_world` there is a check that shiv preserves file permission. Do you remember why does it needs them? Should we keep the logic for now, or I can change the tests?